solr/core/src/java/org/apache/solr/search/neural/KnnQParser.java

```diff
@@ -72,7 +81,49 @@ public Query parse() {
 
     DenseVectorField denseVectorType = (DenseVectorField) fieldType;
     float[] parsedVectorToSearch = parseVector(vectorToSearch, denseVectorType.getDimension());
-    return denseVectorType.getKnnVectorQuery(schemaField.getName(), parsedVectorToSearch, topK);
+
+    return getFilterQuery()
+        .map(
+            filterQuery ->
+                denseVectorType.getKnnVectorQuery(
+                    schemaField.getName(), parsedVectorToSearch, topK, filterQuery))
+        .orElse(
+            denseVectorType.getKnnVectorQuery(schemaField.getName(), parsedVectorToSearch, topK));
+  }
+
+  private Optional<? extends Query> getFilterQuery() throws SolrException {
+    if (!isFilter()) {
+      String[] filterQueries = req.getParams().getParams(CommonParams.FQ);
+      return ofNullable(filterQueries).filter(fq -> fq.length != 0).map(this::computeFilterQuery);
+    }
+    return Optional.empty();
+  }
+
+  private Query computeFilterQuery(String[] filterQueries) throws SolrException {
+
+    List<Query> filters = new ArrayList<>(filterQueries.length);
+
+    for (String filterQuery : filterQueries) {
+      if (filterQuery != null && filterQuery.trim().length() != 0) {
+        QParser filterParser;
+        try {
+          filterParser = QParser.getParser(filterQuery, super.getReq());
```

Comment:
subjective: https://github.com/apache/solr/blob/releases/solr/9.0.0/solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java#L199 and other places have similar code, perhaps a utility method could be factored out somehow to reduce code duplication e.g. in https://github.com/apache/solr/blob/releases/solr/9.0.0/solr/core/src/java/org/apache/solr/search/QueryUtils.java something like `public static List<Query> parseFilterQueries(String[] filterQueries, SolrQueryRequest req)` or so.

Reply:
I did some refactoring in the way you suggested. Now there is a method for parsing filter queries in QueryUtils.java